src/Grpc.AspNetCore.Server/Internal/HttpContextSerializationContext.cs

```diff
@@ -0,0 +1,232 @@
+                GrpcProtocolHelpers.CanWriteCompressed(_serverCallContext.WriteOptions) &&
+                !string.Equals(_serverCallContext.ResponseGrpcEncoding, GrpcProtocolConstants.IdentityGrpcEncoding, StringComparison.Ordinal);
+
+            if (canCompress)
+            {
+                if (_serverCallContext.MethodContext.CompressionProviders.TryGetValue(_serverCallContext.ResponseGrpcEncoding, out var compressionProvider))
+                {
+                    return compressionProvider;
+                }
+            }
+
+            return null;
+        }
+
+        public override void SetPayloadLength(int payloadLength)
+        {
+            switch (_state)
+            {
+                case InternalState.Initialized:
+                    _payloadLength = payloadLength;
+                    break;
+                default:
+                    ThrowInvalidState(_state);
+                    break;
+            }
+        }
+
+        public override void Complete(byte[] payload)
+        {
+            switch (_state)
+            {
+                case InternalState.Initialized:
+                    _state = InternalState.CompleteArray;
+
+                    GrpcServerLog.SerializedMessage(_serverCallContext.Logger, _serverCallContext.MethodContext.ResponseType, payload.Length);
+                    WriteMessage(payload);
+                    break;
+                default:
+                    ThrowInvalidState(_state);
+                    break;
+            }
+        }
+
+        private static void WriteHeader(PipeWriter pipeWriter, int length, bool compress)
+        {
+            const int MessageDelimiterSize = 4; // how many bytes it takes to encode "Message-Length"
+            const int HeaderSize = MessageDelimiterSize + 1; // message length + compression flag
+
+            var headerData = pipeWriter.GetSpan(HeaderSize);
+
+            // Compression flag
+            headerData[0] = compress ? (byte)1 : (byte)0;
+
+            // Message length
+            BinaryPrimitives.WriteUInt32BigEndian(headerData.Slice(1), (uint)length);
+
+            pipeWriter.Advance(HeaderSize);
+        }
+
+        public override IBufferWriter<byte> GetBufferWriter()
+        {
+            switch (_state)
+            {
+                case InternalState.Initialized:
+                    // When writing directly to the buffer the header with message size needs to be written first
+                    if (DirectSerializationSupported)
+                    {
+                        Debug.Assert(_payloadLength != null, "A payload length is required for direct serialization.");
+
+                        EnsureMessageSizeAllowed(_payloadLength.Value);
+
+                        WriteHeader(ResponseBufferWriter, _payloadLength.Value, compress: false);
+                    }
+
+                    _state = InternalState.IncompleteBufferWriter;
+                    goto case InternalState.IncompleteBufferWriter;
```

Comment:
Woah I haven't seen a goto in a long time. Isn't the syntax the same without the goto (i.e. no break)?

Reply:
No. In C# you must have a break.

Reply:
TIL you can only fall through if the case is empty.